File: android/feature/shop/src/main/java/com/enterprise/feature/shop/domain/usecases/CartUseCases.kt
package com.enterprise.feature.shop.domain.usecases

import com.enterprise.core.domain.entities.DomainResult
import com.enterprise.core.domain.repositories.CartRepository
import javax.inject.Inject

/** Removes item from cart. */
class RemoveFromCartUseCase @Inject constructor(
    private val cartRepository: CartRepository
) {
    suspend operator fun invoke(itemId: String): DomainResult<Unit> {
        return cartRepository.removeItem(itemId)
    }
}

/** Updates cart item quantity. */
class UpdateCartQuantityUseCase @Inject constructor(
    private val cartRepository: CartRepository
) {
    suspend operator fun invoke(itemId: String, quantity: Int): DomainResult<Unit> {
        return if (quantity <= 0) {
            cartRepository.removeItem(itemId)
        } else {
            cartRepository.updateQuantity(itemId, quantity)
        }
    }
}

File: ios/EnterpriseMobileApp/Sources/Feature/Shop/Domain/UseCases/CartUseCases.swift
import Foundation
import CoreDomain

/// Use case for removing items from the cart.
public struct RemoveFromCartUseCase: Sendable {
    private let cartRepository: CartRepository

    public init(cartRepository: CartRepository) {
        self.cartRepository = cartRepository
    }

    public func execute(itemId: String) async -> DomainResult<Void> {
        await cartRepository.removeItem(id: itemId)
    }
}

/// Use case for updating item quantity in the cart.
public struct UpdateCartQuantityUseCase: Sendable {
    private let cartRepository: CartRepository

    public init(cartRepository: CartRepository) {
        self.cartRepository = cartRepository
    }

    public func execute(itemId: String, quantity: Int) async -> DomainResult<Void> {
        if quantity <= 0 {
            return await cartRepository.removeItem(id: itemId)
        }
        return await cartRepository.updateQuantity(itemId: itemId, quantity: quantity)
    }
}
